### ISSUE TITLE
PlayCookieSessionStore does not properly work with direct clients

### ISSUE THREAD

**Issue by marcadella** [pac4j/play-pac4j#266]
I use an AnonymousClient to secure an endpoint like this:
```
def anonymousIndex = Secure("AnonymousClient") {
    implicit request: Request[AnyContent] =>
      println(pac4jTemplateHelper.getCurrentProfile.get) //Exception here
      Redirect(routes.HomeController.index())
  }
```
From [this thread](https://groups.google.com/forum/?hl=en#!searchin/pac4j-users/anonymous|sort:date/pac4j-users/FisZbCHcCbk/L8wC8H_HAwAJ) you say that `anonymous` is treated like any authenticated user. So I expect to get `anonymous` in the profile list but instead the Option is empty. I even used `AlwaysUseSessionProfileStorageDecision` to be sure that the profile would be stored.
Here is the log:
```
[debug] o.p.c.e.DefaultSecurityLogic - === SECURITY ===
[debug] o.p.c.e.DefaultSecurityLogic - url: http://localhost:9000/submit
[debug] o.p.c.e.DefaultSecurityLogic - matchers: null
[debug] o.p.c.e.DefaultSecurityLogic - clients: AnonymousClient
[debug] o.p.c.c.f.DefaultSecurityClientFinder - Provided clientNames: AnonymousClient
[debug] o.p.c.c.f.DefaultSecurityClientFinder - clientNameOnRequest: null
[debug] o.p.c.c.f.DefaultSecurityClientFinder - result: [AnonymousClient]
[debug] o.p.c.e.DefaultSecurityLogic - currentClients: [#AnonymousClient# | name: AnonymousClient | credentialsExtractor: org.pac4j.core.client.dir
ect.AnonymousClient$$Lambda$5103/1824587421@55c7ea5f | authenticator: org.pac4j.core.client.direct.AnonymousClient$$Lambda$5104/1651738528@37fe6ca2
 | profileCreator: org.pac4j.core.profile.creator.AuthenticatorProfileCreator@152772e1 | authorizationGenerators: [] | customProperties: {} |]
[debug] o.p.c.e.DefaultSecurityLogic - loadProfilesFromSession: true
[debug] o.p.c.e.DefaultSecurityLogic - profiles: []
[debug] o.p.c.e.DefaultSecurityLogic - Performing authentication for direct client: #AnonymousClient# | name: AnonymousClient | credentialsExtracto
r: org.pac4j.core.client.direct.AnonymousClient$$Lambda$5103/1824587421@55c7ea5f | authenticator: org.pac4j.core.client.direct.AnonymousClient$$Lam
bda$5104/1651738528@37fe6ca2 | profileCreator: org.pac4j.core.profile.creator.AuthenticatorProfileCreator@152772e1 | authorizationGenerators: [] |
customProperties: {} |
[debug] o.p.c.c.d.AnonymousClient - Credentials validation took: 0 ms
[debug] o.p.c.e.DefaultSecurityLogic - credentials: org.pac4j.core.credentials.AnonymousCredentials@0
[debug] o.p.c.c.d.AnonymousClient - credentials : org.pac4j.core.credentials.AnonymousCredentials@0
[debug] o.p.c.c.d.AnonymousClient - profile: #AnonymousProfile# | id: anonymous | attributes: {} | roles: [] | permissions: [] | isRemembered: fals
e | clientName: AnonymousClient | linkedId: null |
[debug] o.p.c.e.DefaultSecurityLogic - profile: #AnonymousProfile# | id: anonymous | attributes: {} | roles: [] | permissions: [] | isRemembered: f
alse | clientName: AnonymousClient | linkedId: null |
[debug] o.p.c.e.DefaultSecurityLogic - saveProfileInSession: true / multiProfile: false
[debug] o.p.c.e.DefaultSecurityLogic - new profiles: [#AnonymousProfile# | id: anonymous | attributes: {} | roles: [] | permissions: [] | isRemembe
red: false | clientName: AnonymousClient | linkedId: null |]
[debug] o.p.c.e.DefaultSecurityLogic - authorizers: null
[debug] o.p.c.e.DefaultSecurityLogic - authenticated and authorized -> grant access
[debug] o.p.p.s.SecureAction$ - profiles: [#AnonymousProfile# | id: anonymous | attributes: {} | roles: [] | permissions: [] | isRemembered: false
| clientName: AnonymousClient | linkedId: null |]
```
In addition, 
> https://github.com/pac4j/play-pac4j-scala-demo, the index page returns an anonymous user and the session remains the same: does it work for you?

I tried it on Heroku and profile is empty.

**Comment by leleuj** [comment]
Please use the mailing list for questions: https://groups.google.com/forum/?fromgroups#!forum/pac4j-users

**Comment by leleuj** [comment]
The problem comes from the PlayCookieSessionStore...

**Comment by marcadella** [comment]
I can confirm that DirectBasicAuthClient doesn't work either.
Do you mean that there is a bug in PlayCookieSessionStore that suddenly broke the library? What's the action plan regarding this issue?

**Comment by leleuj** [comment]
Indeed, there is certainly a bug. It must be investigated and fixed.

**Comment by martinhansen** [comment]
Happens when using PlayCacheSessionStore as well.

From my quick trace/debug, the Secure action finds the profiles from the SessionStore (it even passes the profiles to the AuthenticatedRequest) just fine. But when the block of code is executed from the action (Security.scala line 106), a new play HTTPConfiguration is created (i don't know if this has always been the case), and for some reason, the session id is lost along the way.

I can get it to work in a unit tests by sending a FakeRequest().withSession('foo') when calling a test initialized application.

Using play-pac4j 8.0.1

**Comment by leleuj** [comment]
OK. Thanks for the hint. Maybe it only happens for Scala then. I'll take that into account when I work on it.

**Comment by mikla** [comment]
Looking into `PlayCookieSessionStore.java`
```java
public boolean renewSession(PlayWebContext playWebContext) {
        return false;
    }
```

and then
```java
final boolean renewed = sessionStore.renewSession(context);
            if (renewed) {
              .... 
           } else {
                logger.error("Unable to renew the session. The session store may not support this feature");
            }
```
I wonder, how does it supposed to work?

**Comment by leleuj** [comment]
The session is supposed to be renewed. For example, a new JSESSIONID cookie value is generated to store the session data. In Play, as everything goes in the same cookie, it doesn't make any sense.

**Comment by kcoolsae** [comment]
The same (or similar) problem occurs with indirect clients. 
In my setup - play 2.7.3, AzureADClient - I have 'infinite' redirect problems.

When tracing `PlayCookieSessionStore` I see the user profile being stored in to the session store correctly, and as far as I can see it also ends up in the Http-session (a bit difficult to test, as the play session cookie is encrypted), but after redirect the `DefaultSecurityLogic` no longer finds the user profile in the session store.

**Comment by kcoolsae** [comment]
For what it is worth, the bug mentioned in the comment above turns out not to occur when I use `PlayCacheSessionStore` instead of `PlayCookieSessionStore`. I guess that the user profile information gathered by the `AzureADClient` is too large to store into a cookie and therefore somehow does not end up in the session cookie sent by the redirect.

**Comment by leleuj** [comment]
That might be the explanation. At least, the `PlayCacheSessionStore` works without any issue.

**Comment by leleuj** [comment]
It should be fixed in the latest play-pac4j 9.0.0-RC3-SNAPSHOT version: can you confirm? Thanks
